README.md
[![@nuxt/http](https://http.nuxtjs.org/preview.png)](https://http.nuxtjs.org)

# Nuxt HTTP

[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]
[![Github Actions CI][github-actions-ci-src]][github-actions-ci-href]
[![Codecov][codecov-src]][codecov-href]
[![License][license-src]][license-href]
[![Standard JS][standard-js-src]][standard-js-href]

> HTTP module for Nuxt with a universal way to make HTTP requests to the API backend.

- [✨ &nbsp;Release Notes](https://http.nuxtjs.org/releases)
- [📖 &nbsp;Documentation](https://http.nuxtjs.org)

## Features

- The fluent [ky](https://github.com/sindresorhus/ky) API has been extended with enhancements and shortcuts
- Highly customizable options support for [BaseURL](https://http.nuxtjs.org/options#baseurl)
- Automatically proxy cookies and headers when making requests from server side
- Best practices to avoid token sharing when making server side requests
- Easy proxy support to avoid CORS problems and making deployment easier

[📖 &nbsp;Read more](https://http.nuxtjs.org)

## Development

1. Clone this repository
2. Install dependencies using `yarn install`
3. Start development server using `yarn dev`
4. Then go for instance to `http://localhost:3000/mounted`

## 📖 Contribute to the documentation 

1. Go into the docs folder
2. Install dependencies using `yarn install`
3. Start development server using `yarn dev`

## 📑 License

[MIT License](./LICENSE)

Copyright (c) Nuxt.js Team

<!-- Badges -->
[npm-version-src]: https://flat.badgen.net/npm/v/@nuxt/http
[npm-version-href]: https://npmjs.com/package/@nuxt/http
[npm-downloads-src]: https://flat.badgen.net/npm/dm/@nuxt/http
[npm-downloads-href]: https://npmjs.com/package/@nuxt/http
[github-actions-ci-src]: https://github.com/nuxt/http/workflows/ci/badge.svg
[github-actions-ci-href]: https://github.com/nuxt/http/actions?query=workflow%3Aci
[checks-src]: https://flat.badgen.net/github/checks/nuxt/http/master
[checks-href]: https://github.com/nuxt/http/actions
[codecov-src]: https://flat.badgen.net/codecov/c/github/nuxt/http
[codecov-href]: https://codecov.io/gh/nuxt/http
[standard-js-src]: https://img.shields.io/badge/code_style-standard-brightgreen.svg?style=flat-square
[standard-js-href]: https://standardjs.com
[license-src]: https://img.shields.io/npm/l/@nuxt/http.svg
[license-href]: https://npmjs.com/package/@nuxt/http


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.6.4](https://github.com/nuxt/http/compare/v0.6.3...v0.6.4) (2021-02-27)


### Bug Fixes

* fix import of  AbortController and ReadableStream ([460ac85](https://github.com/nuxt/http/commit/460ac85b777c9b3512060b9597af4b650b0fab1e))

### [0.6.3](https://github.com/nuxt/http/compare/v0.6.2...v0.6.3) (2021-02-27)


### Features

* `$http.getBaseURL` ([#165](https://github.com/nuxt/http/issues/165)) ([332bf34](https://github.com/nuxt/http/commit/332bf34fd5bb977f9c2f3077effc40b0c3334d62))


### Bug Fixes

* use server plugin for polyfill ([74ad5c3](https://github.com/nuxt/http/commit/74ad5c3e0e21934638f055185ddd1772289dcf71))

### [0.6.2](https://github.com/nuxt/http/compare/v0.6.1...v0.6.2) (2020-12-23)


### Bug Fixes

* **pkg:** add missing defu dependency ([9ee0d00](https://github.com/nuxt/http/commit/9ee0d00153190391ac57abd1343e5cc3a29af9b9))

### [0.6.1](https://github.com/nuxt/http/compare/v0.6.0...v0.6.1) (2020-12-02)

## [0.6.0](https://github.com/nuxt/http/compare/v0.5.13...v0.6.0) (2020-10-01)


### ⚠ BREAKING CHANGES

* **deps:** (ky) remove response property for the first argument passed to `beforeRetry` hook. You can access it on `error.response` instead.

### Features

* **deps:** update dependency ky to ^0.24.0 ([#131](https://github.com/nuxt/http/issues/131)) ([6d16a87](https://github.com/nuxt/http/commit/6d16a87d3c94e190312fc292c074be40a28bd1fe))

### [0.5.13](https://github.com/nuxt/http/compare/v0.5.12...v0.5.13) (2020-09-26)


### Bug Fixes

* use regex to avoid falsely replacing ports starting with 80 ([#124](https://github.com/nuxt/http/issues/124)) ([e981e62](https://github.com/nuxt/http/commit/e981e622891a347ad0ba651bf7c48db29a989f84))

### [0.5.12](https://github.com/nuxt/http/compare/v0.5.11...v0.5.12) (2020-09-01)


### Bug Fixes

* **types:** add missing types for `create` and `setBaseURL` ([#122](https://github.com/nuxt/http/issues/122)) ([c002aaa](https://github.com/nuxt/http/commit/c002aaa97c45fa5dfd91325a005fd5cc18a55b57))

### [0.5.11](https://github.com/nuxt/http/compare/v0.5.10...v0.5.11) (2020-08-20)


### Bug Fixes

* **ky-universal:** server importing es bundle ([#121](https://github.com/nuxt/http/issues/121)) ([029dea6](https://github.com/nuxt/http/commit/029dea6f04f82f8df36459a200751af683788302))

### [0.5.10](https://github.com/nuxt/http/compare/v0.5.9...v0.5.10) (2020-07-16)


### Bug Fixes

* **plugin:** debug option broken ([#117](https://github.com/nuxt/http/issues/117)) ([6bc1295](https://github.com/nuxt/http/commit/6bc12959a59b855dc9020e56fa0d0acb7f0e54d0))

### [0.5.9](https://github.com/nuxt/http/compare/v0.5.8...v0.5.9) (2020-07-14)


### Bug Fixes

* set default host ([18e9b11](https://github.com/nuxt/http/commit/18e9b11c395d7ef362ca75dee1ab97c939121864))

### [0.5.8](https://github.com/nuxt/http/compare/v0.5.7...v0.5.8) (2020-07-14)


### Features

* handle runtimeConfig and static target ([b9f5e05](https://github.com/nuxt/http/commit/b9f5e053d393d6a2f695df04d1d3bbc0ac54123f))


### Bug Fixes

* build.transpile should be an array ([823204c](https://github.com/nuxt/http/commit/823204c399149cc149f7a5c31bf93bd6fd729e28)), closes [#115](https://github.com/nuxt/http/issues/115)

### [0.5.7](https://github.com/nuxt/http/compare/v0.5.6...v0.5.7) (2020-06-29)

### [0.5.6](https://github.com/nuxt/http/compare/v0.5.5...v0.5.6) (2020-06-29)


### Bug Fixes

* **plugin:** allow reusing error.response.json ([77b084b](https://github.com/nuxt/http/commit/77b084bc01ae9f95a33ad49a84d99c8765c8725a))

### [0.5.5](https://github.com/nuxt/http/compare/v0.5.4...v0.5.5) (2020-06-17)

### [0.5.4](https://github.com/nuxt/http/compare/v0.5.3...v0.5.4) (2020-06-17)


### Features

* improve `$` helpers and error object ([#113](https://github.com/nuxt/http/issues/113)) ([ada7387](https://github.com/nuxt/http/commit/ada73879f7891fbec712e41188bba70ee829adc2))

### [0.5.3](https://github.com/nuxt/http/compare/v0.5.2...v0.5.3) (2020-06-12)


### Features

* `debug` option ([#109](https://github.com/nuxt/http/issues/109)) ([874b668](https://github.com/nuxt/http/commit/874b668e1e98774e9ad0ef12d402605d00397fb4))
* https detection ([#106](https://github.com/nuxt/http/issues/106)) ([8ae0cd5](https://github.com/nuxt/http/commit/8ae0cd5655752e15d6d5296ef9fed53a843f0fca)), closes [nuxt-community/axios-module#260](https://github.com/nuxt-community/axios-module/issues/260)


### Bug Fixes

* ignore proxy headers `content-md4` and `content-type` ([#107](https://github.com/nuxt/http/issues/107)) ([4152f87](https://github.com/nuxt/http/commit/4152f874a168dcdbceadd92b178b6213b535856b))
* use local version of ky-universal ([#111](https://github.com/nuxt/http/issues/111)) ([e7cb647](https://github.com/nuxt/http/commit/e7cb64704788e38fe0dc685b415b92b90b61df3f))

### [0.5.2](https://github.com/nuxt/http/compare/v0.5.1...v0.5.2) (2020-06-05)


### Features

* allow adding custom headers with nuxt config ([#101](https://github.com/nuxt/http/issues/101)) ([2839260](https://github.com/nuxt/http/commit/2839260ea97b18f259f79bb3ea0e97b0aa7228c3))
* support baseUrl and remove port :443 and :80 when http or https ([#103](https://github.com/nuxt/http/issues/103)) ([fc3e78e](https://github.com/nuxt/http/commit/fc3e78e1307455df506dbc8f6ba5269f28f6e364))

### [0.5.1](https://github.com/nuxt/http/compare/v0.5.0...v0.5.1) (2020-04-29)


### Bug Fixes

* remove leading slash when using prefixUrl ([#96](https://github.com/nuxt/http/issues/96)) ([10fbd67](https://github.com/nuxt/http/commit/10fbd679664c533e8a5fa45afb1f56ac945d8cfa))

## [0.5.0](https://github.com/nuxt/http/compare/v0.4.0...v0.5.0) (2020-04-13)


### New Features

- Allow creating new instances with `$http.create(kyOptions)` (#3) (#94) ([docs](https://http.nuxtjs.org/guide/advanced.html#create-new-instance-based-on-defaults))

lib/module.js
const path = require('path')
const consola = require('consola')

const logger = consola.withScope('nuxt:http')

function httpModule (_moduleOptions) {
  const { nuxt, addPlugin, requireModule } = this

  // Combine options
  const moduleOptions = { ...nuxt.options.http, ..._moduleOptions }

  // Default port
  const defaultPort =
    process.env.API_PORT ||
    moduleOptions.port ||
    process.env.PORT ||
    process.env.npm_package_config_nuxt_port ||
    (nuxt.options.server && nuxt.options.server.port) ||
    3000

  // Default host
  let defaultHost =
    process.env.API_HOST ||
    moduleOptions.host ||
    process.env.HOST ||
    process.env.npm_package_config_nuxt_host ||
    (nuxt.options.server && nuxt.options.server.host) ||
    'localhost'

  /* istanbul ignore if */
  if (defaultHost === '0.0.0.0') {
    defaultHost = 'localhost'
  }

  // Support baseUrl
  if (moduleOptions.baseUrl && !moduleOptions.baseURL) {
    moduleOptions.baseURL = moduleOptions.baseUrl

    delete moduleOptions.baseUrl
  }

  // Default prefix
  const prefix = process.env.API_PREFIX || moduleOptions.prefix || '/'

  // HTTPS
  const https = Boolean(nuxt.options.server && nuxt.options.server.https)

  // Apply defaults
  const options = {
    baseURL: undefined,
    browserBaseURL: undefined,
    debug: false,
    proxyHeaders: true,
    proxyHeadersIgnore: ['accept', 'host', 'cf-ray', 'cf-connecting-ip', 'content-length', 'content-md5', 'content-type'],
    proxy: false,
    retry: false,
    serverTimeout: false,
    clientTimeout: false,
    https,
    headers: {},
    ...moduleOptions
  }
  const toHttps = s => options.https ? s.replace('http://', 'https://') : s

  // ENV overrides
  /* istanbul ignore if */
  if (process.env.API_URL) {
    options.baseURL = process.env.API_URL
  }

  /* istanbul ignore if */
  if (process.env.API_URL_BROWSER) {
    options.browserBaseURL = process.env.API_URL_BROWSER
  }

  // If no baseURL defined, get it from Nuxt server
  if (!options.baseURL) {
    options.baseURL = `http://${defaultHost}:${defaultPort}${prefix}`

    // Update auto generated baseURL after listen for static target as we use random port
    const publicRuntimeConfig = nuxt.options.publicRuntimeConfig = nuxt.options.publicRuntimeConfig || {}
    publicRuntimeConfig.http = publicRuntimeConfig.http || {}

    const privateRuntimeConfig = nuxt.options.privateRuntimeConfig = nuxt.options.privateRuntimeConfig || {}
    privateRuntimeConfig.http = privateRuntimeConfig.http || {}

    // For static exporting
    if (nuxt.options.target === 'static') {
      nuxt.hook('listen', (_, { host = 'localhost', port }) => {
        publicRuntimeConfig.http.browserBaseURL = toHttps(publicRuntimeConfig.http.browserBaseURL || prefix || '/')
        privateRuntimeConfig.http.baseURL = toHttps(privateRuntimeConfig.http.baseURL || `http://${host}:${port}${prefix}`)
      })
    }
  }

  // Normalize options
  if (options.retry === true) {
    options.retry = 2
  } else if (!options.retry) {
    options.retry = 0
  } else if (!isNaN(options.retry)) {
    options.retry = parseInt(options.retry)
  } else if (typeof options.retry === 'object') {
    options.retry = JSON.stringify(options.retry)
  }

  // Default browserBaseURL
  if (typeof options.browserBaseURL === 'undefined') {
    options.browserBaseURL = options.proxy ? prefix : options.baseURL
  }

  // Remove port 443 when https
  if (options.baseURL.includes('https://')) {
    options.baseURL = options.baseURL.replace(':443', '')
  }

  // Remove port 80 when http
  if (/^http:\/\/.*:80(\/|$)/.test(options.baseURL)) {
    options.baseURL = options.baseURL.replace(':80', '')
  }

  // Convert http:// to https:// if https option is on
  options.baseURL = toHttps(options.baseURL)
  options.browserBaseURL = toHttps(options.browserBaseURL)

  // Register plugin
  addPlugin({
    src: path.resolve(__dirname, 'plugin.js'),
    fileName: 'http.js',
    options
  })
  addPlugin({
    src: path.resolve(__dirname, 'plugin.server.js'),
    fileName: 'http.server.js',
    options
  })

  // Proxy integration
  if (options.proxy) {
    requireModule([
      '@nuxtjs/proxy',
      typeof options.proxy === 'object' ? options.proxy : {}
    ])
  }

  // Transpile ky and ky-universal
  nuxt.options.build.transpile = nuxt.options.build.transpile || []
  nuxt.options.build.transpile.push('ky')

  // Set _HTTP_BASE_URL_ for dynamic SSR baseURL
  process.env._HTTP_BASE_URL_ = options.baseURL

  logger.debug(`baseURL: ${options.baseURL}`)
  logger.debug(`browserBaseURL: ${options.browserBaseURL}`)
}

module.exports = httpModule
module.exports.meta = require('../package.json')
